src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,497 @@
+  rw [rhs_apply, coe_nat_nat, ← @legendre_sym.to_jacobi_sym p ⟨pp⟩, qr_sign_symm hqo hpo,
+      qr_sign_neg_one_pow hpo hqo, @quadratic_reciprocity' p q ⟨pp⟩ ⟨pq⟩ hp hq],
+end
+
+/-- The Law of Quadratic Reciprocity for the Jacobi symbol -/
+lemma jacobi_sym_quadratic_reciprocity {a b : ℕ} (ha : odd a) (hb : odd b) :
+  [a | b]ⱼ = (-1) ^ ((a / 2) * (b / 2)) * [b | a]ⱼ :=
+by rw [← qr_sign_neg_one_pow ha hb, qr_sign_symm ha hb, jacobi_sym_quadratic_reciprocity' ha hb]
+
+/-- The Law of Quadratic Reciprocity for the Jacobi symbol: if `a` and `b` are natural numbers
+with `a % 4 = 1` and `b` odd, then `(a / b) = (b / a)`. -/
+theorem jacobi_sym_quadratic_reciprocity_one_mod_four {a b : ℕ} (ha : a % 4 = 1) (hb : odd b) :
+  [a | b]ⱼ = [b | a]ⱼ :=
+by rw [jacobi_sym_quadratic_reciprocity (odd_iff.mpr (odd_of_mod_four_eq_one ha)) hb,
+       pow_mul, neg_one_pow_div_two_of_one_mod_four ha, one_pow, one_mul]
+
+/-- The Law of Quadratic Reciprocityfor the Jacobi symbol: if `a` and `b` are natural numbers
+both congruent to `3` mod `4`, then `(a / b) = -(b / a)`. -/
+theorem jacobi_sym_quadratic_reciprocity_three_mod_four
+  {a b : ℕ} (ha : a % 4 = 3) (hb : b % 4 = 3) :
+  [a | b]ⱼ = - [b | a]ⱼ :=
+let nop := @neg_one_pow_div_two_of_three_mod_four in begin
+  rw [jacobi_sym_quadratic_reciprocity, pow_mul, nop ha, nop hb, neg_one_mul];
+  rwa [odd_iff, odd_of_mod_four_eq_three],
+end
+
+/-- The Jacobi symbol `(a / b)` depends only on `b` mod `4*a` (version for `a : ℕ`). -/
+lemma jacobi_sym_mod_right' (a : ℕ) {b : ℕ} (hb : odd b) : [a | b]ⱼ = [a | b % (4 * a)]ⱼ :=
+begin
+  cases eq_or_ne a 0 with ha₀ ha₀,
+  { rw [ha₀, mul_zero, mod_zero], },
+  have hb' : odd (b % (4 * a)) := odd.mod_even hb (even.mul_right (by norm_num) _),
+  rcases two_pow_mul_odd ha₀ with ⟨e, a', ha₁, ha₂⟩,
+  nth_rewrite 1 [ha₂], nth_rewrite 0 [ha₂],
+  rw [nat.cast_mul, jacobi_sym_mul_left, jacobi_sym_mul_left,
+      jacobi_sym_quadratic_reciprocity' ha₁ hb, jacobi_sym_quadratic_reciprocity' ha₁ hb',
+      nat.cast_pow, jacobi_sym_pow_left, jacobi_sym_pow_left,
+      (by norm_cast : ((2 : ℕ) : ℤ) = 2), jacobi_sym_two hb, jacobi_sym_two hb'],
+  have H : qr_sign b a' * [b | a']ⱼ = qr_sign (b % (4 * a)) a' * [b % (4 * a) | a']ⱼ,
+  { simp_rw [qr_sign],
+    have ha' : (a' : ℤ) ∣ 4 * a,
+    { rw [ha₂, nat.cast_mul, ← mul_assoc],
+      exact dvd_mul_left a' _, },
+    rw [χ₄_nat_mod_four, χ₄_nat_mod_four (b % (4 * a)), mod_mod_of_dvd b (dvd_mul_right 4 a),
+        jacobi_sym_mod_left b, jacobi_sym_mod_left (b % (4 * a)), int.mod_mod_of_dvd b ha'], },
+  cases eq_or_ne e 0 with he he,
+  { rwa [he, pow_zero, pow_zero, one_mul, one_mul], },
+  { have h2 : 8 ∣ 4 * a,
+    { rw [ha₂, ← nat.add_sub_of_le (nat.pos_of_ne_zero he), pow_add, pow_one,
+          ← mul_assoc, ← mul_assoc, (by norm_num : 4 * 2 = 8), mul_assoc],
+      exact dvd_mul_right 8 _, },
+    rw [H, χ₈_nat_mod_eight, χ₈_nat_mod_eight (b % (4 * a)), mod_mod_of_dvd b h2],
+    refl, }
```

Comment:
Golf:
```suggestion
  rcases eq_or_ne a 0 with rfl | ha₀,
  { rw [mul_zero, mod_zero], },
  have hb' : odd (b % (4 * a)) := odd_mod_of_odd hb (even.mul_right (by norm_num) _),
  rcases two_pow_mul_odd ha₀ with ⟨e, a', ha₁, ha₂⟩,
  nth_rewrite 1 [ha₂], nth_rewrite 0 [ha₂],
  rw [nat.cast_mul, jacobi_sym_mul_left, jacobi_sym_mul_left,
      jacobi_sym_quadratic_reciprocity' ha₁ hb, jacobi_sym_quadratic_reciprocity' ha₁ hb',
      nat.cast_pow, jacobi_sym_pow_left, jacobi_sym_pow_left,
      (show ((2 : ℕ) : ℤ) = 2, by refl), jacobi_sym_two hb, jacobi_sym_two hb'],
  congr' 1, swap, congr' 1,
  { simp_rw [qr_sign],
    rw [χ₄_nat_mod_four, χ₄_nat_mod_four (b % (4 * a)), mod_mod_of_dvd b (dvd_mul_right 4 a)] },
  { rw [jacobi_sym_mod_left ↑(b % _), jacobi_sym_mod_left b, int.coe_nat_mod, int.mod_mod_of_dvd b],
    simp only [ha₂, nat.cast_mul, ← mul_assoc],
    exact dvd_mul_left a' _, },
  { cases e, { refl },
    rw [χ₈_nat_mod_eight, χ₈_nat_mod_eight (b % (4 * a)), mod_mod_of_dvd b],
    use 2 ^ e * a', rw [ha₂, pow_succ], ring },
```

Reply:
OK, thanks. At least, it looks like my proofs are no longer that much longer than yours :smile: ...